src/raftstore/store/worker/compact.rs

```diff
@@ -149,4 +318,83 @@ mod test {
             .unwrap();
         assert!(old_sst_files_size > new_sst_files_size);
     }
+
+    fn mvcc_put(db: &DB, k: &[u8], v: &[u8], start_ts: u64, commit_ts: u64) {
+        let cf = get_cf_handle(db, CF_WRITE).unwrap();
+        let k = MvccKey::from_encoded(data_key(k));
+        let k = k.append_ts(commit_ts);
+        let w = Write::new(WriteType::Put, start_ts, Some(v.to_vec()));
+        db.put_cf(cf, k.encoded(), &w.to_bytes()).unwrap();
+    }
+
+    fn delete(db: &DB, k: &[u8], commit_ts: u64) {
+        let cf = get_cf_handle(db, CF_WRITE).unwrap();
+        let k = MvccKey::from_encoded(data_key(k));
+        let k = k.append_ts(commit_ts);
+        db.delete_cf(cf, k.encoded()).unwrap();
+    }
+
+    fn open_db(path: &str) -> DB {
+        let db_opts = rocksdb::DBOptions::new();
+        let mut cf_opts = rocksdb::ColumnFamilyOptions::new();
+        let f = Box::new(MvccPropertiesCollectorFactory::default());
+        cf_opts.add_table_properties_collector_factory("tikv.test-collector", f);
+        let cfs_opts = vec![
+            CFOptions::new(CF_DEFAULT, rocksdb::ColumnFamilyOptions::new()),
+            CFOptions::new(CF_RAFT, rocksdb::ColumnFamilyOptions::new()),
+            CFOptions::new(CF_LOCK, rocksdb::ColumnFamilyOptions::new()),
+            CFOptions::new(CF_WRITE, cf_opts),
+        ];
+        new_engine_opt(path, db_opts, cfs_opts).unwrap()
+    }
+
+    #[test]
+    fn test_check_space_redundancy() {
+        let p = TempDir::new("test").unwrap();
+        let engine = open_db(p.path().to_str().unwrap());
+        let cf = get_cf_handle(&engine, CF_WRITE).unwrap();
+
+        // mvcc_put 0..5
+        for i in 0..5 {
+            let (k, v) = (format!("k{}", i), format!("value{}", i));
+            mvcc_put(&engine, k.as_bytes(), v.as_bytes(), 1, 2);
+        }
+        engine.flush_cf(cf, true).unwrap();
+
+        // gc 0..5
+        for i in 0..5 {
+            let k = format!("k{}", i);
+            delete(&engine, k.as_bytes(), 2);
+        }
+        engine.flush_cf(cf, true).unwrap();
+
+        let (s, e) = (data_key(b"k0"), data_key(b"k5"));
+        let (entries, version) = get_range_entries_and_versions(&engine, cf, &s, &e).unwrap();
+        assert_eq!(entries, 10);
+        assert_eq!(version, 5);
+
+        // mvcc_put 5..10
+        for i in 5..10 {
+            let (k, v) = (format!("k{}", i), format!("value{}", i));
+            mvcc_put(&engine, k.as_bytes(), v.as_bytes(), 1, 2);
+        }
+        engine.flush_cf(cf, true).unwrap();
+
+        let (s, e) = (data_key(b"k5"), data_key(b"k9"));
+        let (entries, version) = get_range_entries_and_versions(&engine, cf, &s, &e).unwrap();
+        assert_eq!(entries, 5);
+        assert_eq!(version, 5);
+
+        let mut ranges_to_check = BTreeSet::new();
```

Comment:
Add more ranges to cover the neighbor ranges case?

Reply:
done